src/job/file_list_dump.rs

```diff
@@ -0,0 +1,425 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use arrow::{
+    array::{BooleanBuilder, Int64Builder, StringBuilder},
+    datatypes::Field,
+    record_batch::RecordBatch,
+};
+use arrow_schema::Schema;
+use chrono::{Datelike, Timelike, Utc};
+use config::{
+    PARQUET_BATCH_SIZE, PARQUET_MAX_ROW_GROUP_SIZE,
+    cluster::{self, LOCAL_NODE},
+    get_config, get_parquet_compression,
+    meta::stream::{FileMeta, StreamType},
+};
+use hashbrown::HashSet;
+use infra::{
+    file_list::FileRecord,
+    schema::{STREAM_SCHEMAS_LATEST, SchemaCache},
+};
+use once_cell::sync::Lazy;
+use parquet::{arrow::AsyncArrowWriter, file::properties::WriterProperties};
+use tokio::sync::{
+    Mutex, RwLock,
+    mpsc::{Receiver, Sender},
+};
+
+use crate::{common::infra::cluster::get_node_by_uuid, service::db};
+
+const HOUR_IN_MS: i64 = 3600 * 1000;
+pub static FILE_LIST_SCHEMA: Lazy<Arc<Schema>> = Lazy::new(|| {
+    Arc::new(Schema::new(vec![
+        Field::new("id", arrow_schema::DataType::Int64, false),
+        Field::new("org", arrow_schema::DataType::Utf8, false),
+        Field::new("stream", arrow_schema::DataType::Utf8, false),
+        Field::new("date", arrow_schema::DataType::Utf8, false),
+        Field::new("file", arrow_schema::DataType::Utf8, false),
+        Field::new("deleted", arrow_schema::DataType::Boolean, false),
+        Field::new("flattened", arrow_schema::DataType::Boolean, false),
+        Field::new("min_ts", arrow_schema::DataType::Int64, false),
+        Field::new("max_ts", arrow_schema::DataType::Int64, false),
+        Field::new("records", arrow_schema::DataType::Int64, false),
+        Field::new("original_size", arrow_schema::DataType::Int64, false),
+        Field::new("compressed_size", arrow_schema::DataType::Int64, false),
+        Field::new("index_size", arrow_schema::DataType::Int64, true),
+    ]))
+});
+pub const FILE_LIST_CACHE_DIR_NAME: &str = "_oo_file_list_dump";
+
+type JobInfo = (i64, String, String, i64);
+
+struct DownloadQueue {
+    sender: Sender<JobInfo>,
+    receiver: Arc<Mutex<Receiver<JobInfo>>>,
+}
+
+impl DownloadQueue {
+    fn new(sender: Sender<JobInfo>, receiver: Arc<Mutex<Receiver<JobInfo>>>) -> Self {
+        Self { sender, receiver }
+    }
+}
+
+const FILE_LIST_DUMP_QUEUE_SIZE: usize = 10000;
+const FILE_LIST_DUMP_THREAD_COUNT: usize = 25;
+static FILE_LIST_DUMP_CHANNEL: Lazy<DownloadQueue> = Lazy::new(|| {
+    let (tx, rx) = tokio::sync::mpsc::channel::<JobInfo>(FILE_LIST_DUMP_QUEUE_SIZE);
+    DownloadQueue::new(tx, Arc::new(Mutex::new(rx)))
+});
+
+static ONGOING_JOB_IDS: Lazy<RwLock<HashSet<i64>>> = Lazy::new(|| RwLock::new(HashSet::new()));
+
+async fn lock_stream(org: &str, stream: &str) -> Option<()> {
```

Comment:
i am thinking maybe we can use the compactor offset for the stream, which node can merge for that stream, then we use same node for dump for the stream. then we don't need this logic.